usr/lib/common/mech_openssl.c

```diff
@@ -5381,3 +5381,189 @@ CK_RV openssl_specific_rsa_prf(CK_BYTE *out, CK_ULONG outlen,
+    }
+
+    return CKR_OK;
+}
+
+CK_RV openssl_specific_ibm_dilithium_generate_keypair(STDLL_TokData_t *tokdata,
+                                                      const struct pqc_oid *oid,
+                                                      TEMPLATE *publ_tmpl,
+                                                      TEMPLATE *priv_tmpl)
+{
+    EVP_PKEY_CTX *ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+    const char *alg_name;
+    CK_BYTE *spki = NULL, *pkcs8 = NULL;
+    CK_ULONG spki_len = 0, pkcs8_len = 0;
+    size_t priv_len = 0, pub_len = 0;
+    CK_BYTE *priv_key = NULL, *pub_key = NULL;
+    CK_RV rc = CKR_OK;
+
+    UNUSED(tokdata);
+
+    alg_name = openssl_get_pqc_oid_name(oid);
+    if (alg_name == NULL) {
+        TRACE_ERROR("Dilithium key form '%lu' not supported by oqsprovider\n",
+                    oid->keyform);
+        rc = CKR_KEY_SIZE_RANGE;
+        goto out;
+    }
+
+    /* Generate key via oqsprovider */
+    ctx = EVP_PKEY_CTX_new_from_name(NULL, alg_name, NULL);
+    if (ctx == NULL) {
+        TRACE_ERROR("EVP_PKEY_CTX_new_from_name failed for '%s'\n", alg_name);
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    if (EVP_PKEY_keygen_init(ctx) != 1) {
+        TRACE_ERROR("EVP_PKEY_keygen_init failed for '%s'\n", alg_name);
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    if (EVP_PKEY_generate(ctx, &pkey) != 1) {
+        TRACE_ERROR("EVP_PKEY_generate failed for '%s'\n", alg_name);
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    /* Get private and public key */
+    rc = get_key_from_pkey(pkey, OSSL_PKEY_PARAM_PRIV_KEY,
+                           &priv_key, &priv_len);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("get_key_from_pkey failed for priv key\n");
+        goto out;
+    }
+
+    rc = get_key_from_pkey(pkey, OSSL_PKEY_PARAM_PUB_KEY,
+                           &pub_key, &pub_len);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("get_key_from_pkey failed for pub key\n");
+        goto out;
+    }
+
+    /* Extract key components */
+    rc = ibm_dilithium_unpack_priv_key(priv_key, priv_len, oid, priv_tmpl);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("ibm_dilithium_unpack_priv_key failed for priv key\n");
+        goto out;
+    }
+
+    rc = ibm_dilithium_unpack_pub_key(pub_key, pub_len, oid, publ_tmpl);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("ibm_dilithium_unpack_priv_key failed for pub key\n");
+        goto out;
+    }
+
+    /* Also add public key components to private template */
+    rc = ibm_dilithium_unpack_pub_key(pub_key, pub_len, oid, priv_tmpl);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("ibm_dilithium_unpack_priv_key failed for pub key\n");
+        goto out;
+    }
+
+    /* Add keyform and mode attributes to public and private template */
+    rc = ibm_pqc_add_keyform_mode(publ_tmpl, oid, CKM_IBM_DILITHIUM);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("ibm_pqc_add_keyform_mode failed\n");
```

Comment:
Don't know if the rc could be interesting in the trace msg ... also below

Reply:
ibm_pqc_add_keyform_mode prints its own trace messages in case of a failure, so I guess printing the return code is not so interesting here.